nbserverproxy/handlers.py

```diff
@@ -5,18 +5,88 @@
-    proxy_uri = "http://localhost"
+# from https://stackoverflow.com/questions/38663666/how-can-i-serve-a-http-page-and-a-websocket-on-the-same-url-in-tornado
+class WebSocketHandlerMixin(websocket.WebSocketHandler):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # since my parent doesn't keep calling the super() constructor,
+        # I need to do it myself
+        bases = type(self).__bases__
+        assert WebSocketHandlerMixin in bases
+        meindex = bases.index(WebSocketHandlerMixin)
+        try:
+            nextparent = bases[meindex + 1]
+        except IndexError:
+            raise Exception("WebSocketHandlerMixin should be followed "
+                            "by another parent to make sense")
+
+        # undisallow methods --- t.ws.WebSocketHandler disallows methods,
+        # we need to re-enable these methods
+        def wrapper(method):
+            def undisallow(*args2, **kwargs2):
+                getattr(nextparent, method)(self, *args2, **kwargs2)
+            return undisallow
+
+        for method in ["write", "redirect", "set_header", "set_cookie",
+                       "set_status", "flush", "finish"]:
+            setattr(self, method, wrapper(method))
+        nextparent.__init__(self, *args, **kwargs)
+
+    async def get(self, *args, **kwargs):
+        if self.request.headers.get("Upgrade", "").lower() != 'websocket':
+            return await self.http_get(*args, **kwargs)
+        # super get is not async
+        super().get(*args, **kwargs)
+
+class LocalProxyHandler(WebSocketHandlerMixin, IPythonHandler):
+    async def open(self, port, proxied_path):
+        """
+        Called when a client opens a websocket connection.
+
+        We establish a websocket connection to the proxied backend &
+        set up a callback to relay messages through.
+        """
+        client_uri = '{uri}:{port}{path}'.format(
+            uri='ws://localhost',
+            port=port,
+            path=proxied_path
+        )
+        if self.request.query:
+            client_uri += '?' + self.request.query
+
+        def cb(message):
+            """
+            Callback when the backend sends messages to us
+
+            We just pass it back to the frontend
+            """
+            self.write_message(message)
+        self.ws = await websocket.websocket_connect(client_uri, on_message_callback=cb)
+
+    async def on_message(self, message):
+        """
+        Called when we receive a message from our client.
+
+        We proxy it to the backend.
+        """
+        await self.ws.write_message(message)
+
+    async def on_close(self):
+        """
+        Called when the client closes our websocket connection.
+
+        We close our connection to the backend too.
+        """
+        self.ws.close()
 
     @web.authenticated
-    @gen.coroutine
-    def proxy(self, port, proxied_path):
+    async def proxy(self, port, proxied_path):
```

Comment:
what does this change to do the minimum python version required?

Reply:
makes it Python 3.5

Reply:
I put that info in the commit message, forgot to put it in PR. Have done so now!